File: apps/app/src/services/ai/prompts/yaml-loader.ts
/*
  Cross-language YAML prompt loader (TypeScript)
  - Requires: npm i yaml
  - Respects PROMPTS_DIR or defaults to repoRoot/prompts
*/
import fs from 'node:fs/promises';
import path from 'node:path';
import YAML from 'yaml';

export type LoadedPrompt = {
  id: string;
  version?: string;
  prompt: string;
  schema?: any;
  meta: Record<string, any>;
};

function interpolate(template: string, params: Record<string, any> = {}): string {
  return template.replace(/\{\{\s*([a-zA-Z0-9_]+)\s*\}\}/g, (_m, key) => {
    const v = params[key];
    return v === undefined || v === null ? '' : String(v);
  });
}

async function fileExists(p: string) {
  try { await fs.access(p); return true; } catch { return false; }
}

async function resolvePromptsDir(custom?: string): Promise<string> {
  if (custom) return custom;
  if (process.env.PROMPTS_DIR) return process.env.PROMPTS_DIR;
  const cwd = process.cwd();
  const candidates = [
    // Prefer new monorepo location
    path.join(cwd, 'packages', 'prompts'),
    // Legacy root-level prompts folder
    path.join(cwd, 'prompts'),
    // Resolve relative to this file back to repo root
    path.resolve(__dirname, '../../../../..', 'packages', 'prompts'),
    path.resolve(__dirname, '../../../../..', 'prompts'),
  ];
  for (const p of candidates) {
    if (await fileExists(p)) return p;
  }
  // Fallback to last candidate if none exist
  return candidates[candidates.length - 1];
}

export async function listPrompts(options?: { promptsDir?: string }) {
  const dir = await resolvePromptsDir(options?.promptsDir);
  let indexPath = path.join(dir, 'registry.yaml');
  if (!(await fileExists(indexPath))) {
    indexPath = path.join(dir, 'index.yaml');
  }
  const raw = await fs.readFile(indexPath, 'utf8');
  const parsed = YAML.parse(raw) as { prompts: Array<Record<string, any>> };
  return parsed.prompts as Array<{ id: string; file: string; schema?: string; version?: string; tags?: string[]; aliases?: string[] }>;
}

export async function loadPrompt(
  id: string,
  params?: Record<string, any>,
  options?: { promptsDir?: string }
): Promise<LoadedPrompt> {
  const dir = await resolvePromptsDir(options?.promptsDir);
  const idx = await listPrompts({ promptsDir: dir });
  const entry = idx.find(p => p.id === id);
  if (!entry) throw new Error(`Prompt not found in index: ${id}`);

  const tplPath = path.join(dir, entry.file);
  const tplRaw = await fs.readFile(tplPath, 'utf8');
  const tpl = YAML.parse(tplRaw) as any;

  const rendered = interpolate(String(tpl.prompt ?? ''), params);

  let schema: any | undefined;
  const schemaRef = tpl.schema_ref ?? entry.schema; // allow either
  if (schemaRef) {
    const schemaPath = path.join(dir, schemaRef);
    if (await fileExists(schemaPath)) {
      const sRaw = await fs.readFile(schemaPath, 'utf8');
      try { schema = JSON.parse(sRaw); } catch { schema = undefined; }
    }
  }

  const meta = {
    id: tpl.id ?? id,
    version: tpl.version ?? entry.version,
    description: tpl.description,
    owner: tpl.owner,
    tags: tpl.tags,
    runtime: tpl.runtime,
    variables: tpl.variables,
    schema_ref: schemaRef,
    source: path.relative(process.cwd(), tplPath),
  };

  return { id, version: meta.version, prompt: rendered, schema, meta };
}

File: packages/ai/prompts/yaml_loader.py
"""
Cross-language YAML prompt loader (Python)
- Requires: pip install pyyaml
- Respects PROMPTS_DIR or defaults to repo_root/prompts
"""
from __future__ import annotations
import os
import json
import re
from pathlib import Path
from typing import Any, Dict, Optional

try:
    import yaml  # type: ignore
except Exception as e:  # pragma: no cover
    raise RuntimeError("pyyaml is required: pip install pyyaml") from e


def _interpolate(template: str, params: Optional[Dict[str, Any]] = None) -> str:
    params = params or {}
    pattern = re.compile(r"\{\{\s*([a-zA-Z0-9_]+)\s*\}\}")

    def _repl(match: re.Match[str]) -> str:
        key = match.group(1)
        val = params.get(key)
        return "" if val is None else str(val)

    return pattern.sub(_repl, template)


def _resolve_prompts_dir(custom: Optional[str] = None) -> Path:
    if custom:
        return Path(custom)
    env = os.getenv("PROMPTS_DIR")
    if env:
        return Path(env)

    # This file lives in packages/ai/prompts — prefer that first
    here = Path(__file__).resolve().parent
    if (here / "registry.yaml").exists():
        return here

    cwd = Path.cwd()
    candidates = [
        cwd / "packages" / "ai" / "prompts",
        cwd / "packages" / "prompts",  # legacy
        cwd / "prompts",
    ]
    for cand in candidates:
        if cand.exists() and (cand / "registry.yaml").exists():
            return cand

    # Ascend to locate packages/ai/prompts (or legacy packages/prompts)
    for base in list(Path(__file__).resolve().parents):
        for rel in ("packages/ai/prompts", "packages/prompts", "prompts"):
            pkg = base / rel
            if pkg.exists() and (pkg / "registry.yaml").exists():
                return pkg

    # Last resort: packages/ai/prompts under CWD (may not exist)
    return cwd / "packages" / "ai" / "prompts"


def _find_registry_entry(index: list, prompt_id: str) -> Optional[Dict[str, Any]]:
    for entry in index:
        if entry.get("id") == prompt_id:
            return entry
        aliases = entry.get("aliases") or []
        if prompt_id in aliases:
            return entry
    return None


def list_prompts(prompts_dir: Optional[str] = None) -> Any:
    base = _resolve_prompts_dir(prompts_dir)
    index_path = base / "registry.yaml"
    if not index_path.exists():
        raise FileNotFoundError(f"Prompt registry not found: {index_path}")
    data = yaml.safe_load(index_path.read_text())
    return data.get("prompts", [])


def load_prompt(id: str, params: Optional[Dict[str, Any]] = None, prompts_dir: Optional[str] = None) -> Dict[str, Any]:
    base = _resolve_prompts_dir(prompts_dir)
    index = list_prompts(str(base))
    entry = _find_registry_entry(index, id)
    if not entry:
        raise FileNotFoundError(f"Prompt not found in registry: {id}")

    tpl_path = base / entry.get("file")
    tpl = yaml.safe_load(tpl_path.read_text())

    rendered = _interpolate(str(tpl.get("prompt", "")), params)

    schema = None
    schema_ref = tpl.get("schema_ref") or entry.get("schema")
    if schema_ref:
        schema_path = base / schema_ref
        if schema_path.exists():
            try:
                schema = json.loads(schema_path.read_text())
            except Exception:
                schema = None

    try:
        source = str(tpl_path.relative_to(Path.cwd()))
    except ValueError:
        source = str(tpl_path)

    meta = {
        "id": tpl.get("id", id),
        "version": tpl.get("version", entry.get("version")),
        "description": tpl.get("description"),
        "owner": tpl.get("owner"),
        "tags": tpl.get("tags"),
        "runtime": tpl.get("runtime"),
        "variables": tpl.get("variables"),
        "schema_ref": schema_ref,
        "source": source,
    }

    return {"id": id, "version": meta.get("version"), "prompt": rendered, "schema": schema, "meta": meta}
